server/target/target.go

```diff
@@ -184,8 +151,129 @@ func readTargets(ctx context.Context, env environment.Env, req *trpb.GetTargetRe
 			Target:       target,
 			TargetStatus: statuses[targetID],
-			RepoUrl:      tq.GetRepoUrl(),
+			RepoUrl:      repoURL,
 		})
 	}
-	return targetHistories, nil
+	resp := &trpb.GetTargetResponse{
+		InvocationTargets: targetHistories,
+	}
+	if paginationToken != nil {
+		tokenStr, err := paginationToken.Encode()
+		if err != nil {
+			return nil, err
+		}
+		resp.NextPageToken = tokenStr
+	}
+	return resp, nil
+}
+
+func readPaginatedTargets(ctx context.Context, env environment.Env, req *trpb.GetTargetRequest) (*trpb.GetTargetResponse, error) {
+	if env.GetDBHandle() == nil {
+		return nil, status.FailedPreconditionError("database not configured")
+	}
+
+	repo := req.GetQuery().GetRepoUrl()
+	if repo != "" {
+		if norm, err := gitutil.NormalizeRepoURL(repo); err == nil {
+			repo = norm.String()
+		} else {
+			return nil, status.InvalidArgumentErrorf("Invalid repo_url: %q", repo)
+		}
+	}
+
+	// Build the query to select the distinct commits.
+	commitQuery := query_builder.NewQuery(`
+		SELECT commit_sha, max(created_at_usec) as latest_created_at_usec 
+		FROM Invocations `)
+	commitQuery.AddWhereClause("group_id = ?", req.GetRequestContext().GetGroupId())
+	if repo != "" {
+		commitQuery.AddWhereClause("repo_url = ?", repo)
+	}
+	commitQuery.AddWhereClause("role = ?", ciRole)
+	paginationToken, err := NewTokenFromRequest(req)
+	if err != nil {
+		return nil, err
+	}
+	if paginationToken != nil {
+		paginationToken.ApplyToQuery(commitQuery)
+	}
+
+	commitQuery.SetGroupBy("commit_sha")
+	commitQuery.SetOrderBy("latest_created_at_usec", false /*=ascending*/)
+	commitQuery.SetLimit(targetPageSize)
+
+	// Build the subquery to select columns from Invocations Table
+	joinQuery := query_builder.NewQuery(`
+		SELECT invocation_uuid, invocation_id, inv.commit_sha, branch_name, repo_url, 
+		created_at_usec 
+		FROM Invocations as inv`)
+	joinQuery.AddJoinClause(commitQuery, "commits", "inv.commit_sha=commits.commit_sha")
+	joinQuery.AddWhereClause("inv.group_id = ?", req.GetRequestContext().GetGroupId())
+	joinQuery.AddWhereClause("inv.role = ?", ciRole)
+	if repo != "" {
+		joinQuery.AddWhereClause("inv.repo_url = ?", repo)
+	}
+	if err := perms.AddPermissionsCheckToQueryWithTableAlias(ctx, env, joinQuery, "inv"); err != nil {
+		return nil, err
+	}
+
+	// Build the full query to get columns from Targets, TargetStatuses and Invocations
+	// Table.
+	q := query_builder.NewQuery(`
+		SELECT t.target_id, t.label, t.rule_type, ts.target_type, ts.test_size, ts.status,
+		ts.start_time_usec, ts.duration_usec, i.invocation_id, i.commit_sha, i.branch_name,
+		i.repo_url, i.created_at_usec
+		FROM Targets as t
+		JOIN TargetStatuses as ts ON ts.target_id = t.target_id`)
+	q.AddJoinClause(joinQuery, "i", "ts.invocation_uuid = i.invocation_uuid")
+	return fetchTargetsFromDB(ctx, env, q, repo)
+}
+
+func readTargets(ctx context.Context, env environment.Env, req *trpb.GetTargetRequest, startUsec, endUsec int64) (*trpb.GetTargetResponse, error) {
```

Comment:
Can you move this above `fetchTargetsFromDB`? I think that will reduce the diff and make the PR a bit easier to review

Reply:
Done